### PR TITLE
Add management command to notify and delete users

### PR DESCRIPTION
**Pull request by kevinetienne**

### CODE REVIEW

**Comment by kevinetienne** [comment]
@Ian-Foote @adam-thomas @NoLogo can you review please?

File: tests/test_management_commands.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+from .factories import UserFactory
+
+
+MONTH = 1
+
+
+class TestUserNotifyManagementCommand(TestCase):
+    def test_no_users(self):
+        call_command('notify_users')
+        self.assertFalse(len(mail.outbox))
```
Comment (LilyFirefly) [inline tests/test_management_commands.py]: I'd prefer `assertEqual(0)`.

File: tests/test_management_commands.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        call_command('delete_users')
+        self.assertFalse(len(mail.outbox))
+
+    def test_inactive_users(self):
+        year_ago = timezone.now() - relativedelta(months=13)
+        UserFactory.create(last_login=year_ago, notified=True)
+
+        call_command('delete_users')
+
+        self.assertEqual(len(mail.outbox), 1)
```
Comment (adam-thomas) [inline tests/test_management_commands.py]: Should we assert that the user has actually been deleted here?  Something like:

```
user = UserFactory.create(...)

call_command('delete_users')

with self.assertRaises(SomeExceptionICantRemember):
    user.refresh_from_db()
```

File: tests/test_management_commands.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+MONTH = 1
+
+
+class TestUserNotifyManagementCommand(TestCase):
+    def test_no_users(self):
+        call_command('notify_users')
+        self.assertFalse(len(mail.outbox))
+
+    def test_inactive_users(self):
+        month_ago = timezone.now() - relativedelta(months=13)
```
Comment (LilyFirefly) [inline tests/test_management_commands.py]: I'm not sure `month_ago` is the best name, but I'm not sure what to propose instead.

Comment (LilyFirefly) [inline tests/test_management_commands.py]: Maybe just `last_login`.

File: tests/test_management_commands.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        self.assertEqual(len(mail.outbox), 1)
+
+    def test_email(self):
+        year_ago = timezone.now() - relativedelta(months=13)
+        UserFactory.create(last_login=year_ago, notified=True)
+        call_command('delete_users')
+
+        email = mail.outbox[0]
+        self.assertEqual(email.subject, 'Your account has been deleted')
+        self.assertIn('You were informed', email.body)
```
Comment (adam-thomas) [inline tests/test_management_commands.py]: Most of this test duplicates `test_inactive_users`, which is both un-dry and potentially slow.  Can we either call the email-generating method directly, or slot the last two asserts into `test_inactive_users`?

Comment (adam-thomas) [inline tests/test_management_commands.py]: For clarity:

```
   def test_inactive_users(self):
        year_ago = timezone.now() - relativedelta(months=13)
        UserFactory.create(last_login=year_ago, notified=True)

        call_command('delete_users')
        self.assertEqual(len(mail.outbox), 1)

        email = mail.outbox[0]
        self.assertEqual(email.subject, 'Your account has been deleted')
        self.assertIn('You were informed', email.body)
```

Comment (adam-thomas) [inline tests/test_management_commands.py]: (We can also assert that the user is gone, as I commented above.)

File: tests/test_management_commands.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        email = mail.outbox[0]
+        self.assertEqual(email.subject, 'Re-activate your account')
+        self.assertIn('We have noticed', email.body)
+
+    def test_notified_users(self):
+        year_ago = timezone.now() - relativedelta(months=13)
+        UserFactory.create(last_login=year_ago, notified=True)
+
+        call_command('notify_users')
+        self.assertFalse(len(mail.outbox))
```
Comment (LilyFirefly) [inline tests/test_management_commands.py]: Can we have consistent spacing? I'd prefer a line before and after `call_command`.

File: tests/test_management_commands.py (hunk 1 above)
Comment (adam-thomas) [inline tests/test_management_commands.py]: :+1: for this, personally.  Not a dealbreaker for me though.

File: tests/test_management_commands.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    def test_inactive_users_config(self):
+        month_ago = timezone.now() - relativedelta(months=MONTH + 1)
+        UserFactory.create(last_login=month_ago)
+
+        call_command('notify_users')
+
+        self.assertEqual(len(mail.outbox), 1)
+
+    def test_email(self):
+        year_ago = timezone.now() - relativedelta(months=12)
```
Comment (LilyFirefly) [inline tests/test_management_commands.py]: Why `12` here but `13` in `test_inactive_users`?

File: user_deletion/management/commands/delete_users.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+User = get_user_model()
+
+
+class Command(BaseCommand):
+    def handle(self, *args, **options):
+        translation.activate(settings.LANGUAGE_CODE)
+        users = User.objects.users_to_delete()
+        site = Site.objects.get_current()
+        DeletedNotification(user=None, site=site, users=users).notify()
```
Comment (adam-thomas) [inline user_deletion/management/commands/delete_users.py]: Wait.  This doesn't actually delete any users.

File: tests/test_management_commands.py (hunk 3 above)
Comment (LilyFirefly) [inline tests/test_management_commands.py]: Why is `13` months significant? Magic numbers are bad. :wink:

File: tests/test_managers.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        user = UserFactory.create(last_login=last_login)
+        users = User.objects.users_to_notify()
+
+        self.assertCountEqual(users, [user])
+
+    def test_users_not_to_notify(self):
+        UserFactory.create(last_login=timezone.now())
+        users = User.objects.users_to_notify()
+
+        self.assertFalse(users)
```
Comment (LilyFirefly) [inline tests/test_managers.py]: I'd prefer the more explicit `self.assertNotIn(user, users)`.

File: tests/test_managers.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+user_deletion_config = apps.get_app_config('user_deletion')
+
+
+class TestUserDeletionManager(TestCase):
+    def test_users_to_notify(self):
+        last_login = timezone.now() - relativedelta(
+            months=user_deletion_config.MONTH_NOTIFICATION,
+        )
+        user = UserFactory.create(last_login=last_login)
```
Comment (LilyFirefly) [inline tests/test_managers.py]: Perhaps make `notified=False` explicit here?

File: user_deletion/apps.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+from django.apps import AppConfig
+
+
+class UserDeletionConfig(AppConfig):
+    name = 'user_deletion'
+
+    # users are notificed after 12 months of inactivity
+    MONTH_NOTIFICATION = 12
+    # users are deleted after 13 months
+    MONTH_DELETION = 13
```
Comment (LilyFirefly) [inline user_deletion/apps.py:10]: What if someone wants to specify a number of days or weeks instead?

File: user_deletion/managers.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
+from dateutil.relativedelta import relativedelta
+from django.apps import apps
+from django.utils import timezone
+
+
+user_deletion_config = apps.get_app_config('user_deletion')
+
+
+class UserManagerMixin:
```
Comment (LilyFirefly) [inline user_deletion/managers.py]: Can we have a better name? This will be meaningless in a project.

File: user_deletion/managers.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    def users_to_notify(self):
+        """Finds all users who have been inactive and not yet notified."""
+        last_login = timezone.now() - relativedelta(
+            months=user_deletion_config.MONTH_NOTIFICATION,
+        )
```
Comment (LilyFirefly) [inline user_deletion/managers.py:14]: This isn't `last_login`, it's the notification boundary datetime.

File: user_deletion/notifications.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    messages = []
+    context = {'site': notification.site}
+    for user in notification.users:
+        message = render_to_string(notification.template_name, context)
+        messages.append([
+            notification.subject,
+            message,
+            settings.DEFAULT_FROM_EMAIL,
+            [user.email],
+        ])
```
Comment (LilyFirefly) [inline user_deletion/notifications.py]: I'd be tempted to use a generator instead of `messages` and `append`.

File: user_deletion/apps.py (hunk 10 above)
Comment (adam-thomas) [inline user_deletion/apps.py:10]: That would probably be a separate PR, right?  Not a difficult change conceptually, but enough of a faff that I think we can do it elsewhere.

File: user_deletion/notifications.py [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    send_mass_mail(messages)
+
+
+class DeletionNotification(Notification):
+    handlers = (send_emails,)
+    template_name = 'user_deletion/email_notification.txt'
+    subject = _('Re-activate your account')
+
+
+class DeletedNotification(Notification):
```
Comment (LilyFirefly) [inline user_deletion/notifications.py]: These two names are easily confused. If we need to be more verbose for clarity, so be it.

File: user_deletion/managers.py (hunk 11 above)
Comment (adam-thomas) [inline user_deletion/managers.py]: :+1: this for definite

File: tests/test_management_commands.py (hunk 2 above)
Comment (LilyFirefly) [inline tests/test_management_commands.py]: `User.DoesNotExist` I think.

Comment (adam-thomas) [inline tests/test_management_commands.py]: Probably, thanks :)

File: user_deletion/management/commands/delete_users.py (hunk 7 above)
Comment (kevinetienne) [inline user_deletion/management/commands/delete_users.py]: Haha :) Thanks

Comment (adam-thomas) [inline user_deletion/management/commands/delete_users.py]: Seems like it could be a problem :P

File: user_deletion/apps.py (hunk 10 above)
Comment (kevinetienne) [inline user_deletion/apps.py:10]: Added #3

File: user_deletion/managers.py (hunk 12 above)
Comment (kevinetienne) [inline user_deletion/managers.py:14]: Do you have a suggestion I'm not sure what to replace it with?

File: user_deletion/apps.py (hunk 10 above)
Comment (adam-thomas) [inline user_deletion/apps.py:10]: :+1:

File: user_deletion/managers.py (hunk 12 above)
Comment (LilyFirefly) [inline user_deletion/managers.py:14]: `boundary`? `notification_boundary`? `inactive_boundary`? It's not a very easy one to name I agree.

Comment (kevinetienne) [inline user_deletion/managers.py:14]: Cool cheers

File: tests/test_management_commands.py (hunk 4 above)
Comment (kevinetienne) [inline tests/test_management_commands.py]: That's a good point I'll add test for the notifications instead

File: user_deletion/managers.py (hunk 12 above)
Comment (adam-thomas) [inline user_deletion/managers.py:14]: `threshold` is another potential term to work with.

Comment (adam-thomas) [inline user_deletion/managers.py:14]: http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=29785&type=card

Comment (kevinetienne) [inline user_deletion/managers.py:14]: updated in https://github.com/incuna/django-user-deletion/pull/2/commits/7ce51c694e44e8503acd86de0f90dbc4078f4b82

**Comment by KevinEtienne** [comment]
Updated

**Comment by kevinetienne** [comment]
Updated and thanks for the review!

**Comment by adam-thomas** [comment]
LGTM, nice one! :)

It occurs to me that the 'account deleted' email sounds quite hostile.  Where does that copy come from?

**Comment by kevinetienne** [comment]
It comes from [diabates coach](https://github.com/incuna/az-diabetescoach-epatient/blob/master/api/users/templates/users/deletion_email.txt)

**Comment by adam-thomas** [comment]
OK cool, fine by me then :)
